sui-execution/latest/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -357,16 +358,22 @@ impl<'vm, 'state, 'a> ExecutionContext<'vm, 'state, 'a> {
         {
             return Err(CommandArgumentError::InvalidGasCoinUsage);
         }
-        // Immutable objects and shared objects cannot be taken by value
-        if matches!(
-            input_metadata_opt,
-            Some(InputObjectMetadata {
-                owner: Owner::Immutable | Owner::Shared { .. },
-                ..
-            })
-        ) {
-            return Err(CommandArgumentError::InvalidObjectByValue);
+
+        match input_metadata_opt {
+            Some(InputObjectMetadata { owner: Owner::Immutable, .. }) => {
+                return Err(CommandArgumentError::InvalidObjectByValue)
+            }
+            Some(InputObjectMetadata { owner: Owner::Shared {..}, .. })
+            if !protocol_config.shared_object_deletion() => {
+                return Err(CommandArgumentError::InvalidObjectByValue)
+            }
+            Some(InputObjectMetadata { owner: Owner::Shared {..}, .. })
+            if protocol_config.shared_object_deletion() && matches!(command_kind, CommandKind::TransferObjects) => {
```

Comment:
The second match seems error-prone to me. Are sure sure all other commands can accept shared object by value?
If we only want to allow it in MoveCall, we should probably make that explicit.

Reply:
This is true -- I don't think we want to support people merging shared coins?